core/src/main/java/org/apache/calcite/sql/fun/SqlDialectOperatorTableFactory.java

```diff
@@ -0,0 +1,247 @@
+      new HashMap<>();
+
+  /** Return type inference for {@code DECODE}. */
+  protected static final SqlReturnTypeInference DECODE_RETURN_TYPE =
+      opBinding -> {
+        final List<RelDataType> list = new ArrayList<>();
+        for (int i = 1, n = opBinding.getOperandCount(); i < n; i++) {
+          if (i < n - 1) {
+            ++i;
+          }
+          list.add(opBinding.getOperandType(i));
+        }
+        final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        RelDataType type = typeFactory.leastRestrictive(list);
+        if (opBinding.getOperandCount() % 2 == 1) {
+          type = typeFactory.createTypeWithNullability(type, true);
+        }
+        return type;
+      };
+
+  /** The "DECODE(v, v1, result1, [v2, result2, ...], resultN)" function. */
+  @SqlDialect(dialects = {Dialect.ORACLE})
+  public static final SqlFunction DECODE =
+      new SqlFunction("DECODE", SqlKind.DECODE, DECODE_RETURN_TYPE, null,
+          OperandTypes.VARIADIC, SqlFunctionCategory.SYSTEM);
+
+  /** The "NVL(value, value)" function. */
+  @SqlDialect(dialects = {Dialect.ORACLE})
+  public static final SqlFunction NVL =
+      new SqlFunction("NVL", SqlKind.NVL,
+          ReturnTypes.cascade(ReturnTypes.LEAST_RESTRICTIVE,
+              SqlTypeTransforms.TO_NULLABLE_ALL),
+          null, OperandTypes.SAME_SAME, SqlFunctionCategory.SYSTEM);
+
+  /** The "LTRIM(string)" function. */
+  @SqlDialect(dialects = {Dialect.ORACLE})
+  public static final SqlFunction LTRIM =
+      new SqlFunction("LTRIM", SqlKind.LTRIM,
+          ReturnTypes.cascade(ReturnTypes.ARG0, SqlTypeTransforms.TO_NULLABLE,
+              SqlTypeTransforms.TO_VARYING), null,
+          OperandTypes.STRING, SqlFunctionCategory.STRING);
+
+  /** The "RTRIM(string)" function. */
+  @SqlDialect(dialects = {Dialect.ORACLE})
+  public static final SqlFunction RTRIM =
+      new SqlFunction("RTRIM", SqlKind.RTRIM,
+          ReturnTypes.cascade(ReturnTypes.ARG0, SqlTypeTransforms.TO_NULLABLE,
+              SqlTypeTransforms.TO_VARYING), null,
+          OperandTypes.STRING, SqlFunctionCategory.STRING);
+
+  /** Oracle's "SUBSTR(string, position [, substringLength ])" function.
+   *
+   * <p>It has similar semantics to standard SQL's
+   * {@link SqlStdOperatorTable#SUBSTRING} function but different syntax. */
+  @SqlDialect(dialects = {Dialect.ORACLE})
+  public static final SqlFunction SUBSTR =
+      new SqlFunction("SUBSTR", SqlKind.OTHER_FUNCTION,
+          ReturnTypes.ARG0_NULLABLE_VARYING, null, null,
+          SqlFunctionCategory.STRING);
+
+  /** The "GREATEST(value, value)" function. */
+  @SqlDialect(dialects = {Dialect.ORACLE})
+  public static final SqlFunction GREATEST =
+      new SqlFunction("GREATEST", SqlKind.GREATEST,
+          ReturnTypes.cascade(ReturnTypes.LEAST_RESTRICTIVE,
+              SqlTypeTransforms.TO_NULLABLE), null,
+          OperandTypes.SAME_VARIADIC, SqlFunctionCategory.SYSTEM);
+
```

Comment:
What if both Oracle and MySql have the function but with different syntax?

Reply:
Then i think we should keep a different field name for them, after all, these two sql function should never be put into one sql parse context